sdk/granter.go

```diff
@@ -323,8 +323,26 @@ func newGranterFromAttributes(attrs ...*policy.Value) (granter, error) {
 		if def == nil {
 			return granter{}, fmt.Errorf("no associated definition with value [%s]", fqn)
 		}
-		grants.addAllGrants(fqn, def.GetGrants(), def)
-		grants.addAllGrants(fqn, v.GetGrants(), def)
+		namespace := def.GetNamespace()
+		if namespace == nil {
+			return granter{}, fmt.Errorf("no associated namespace with definition [%s] from value [%s]", def.GetFqn(), fqn)
+		}
+
+		valuesGranted := false
+		attributesGranted := false
+		if v != nil {
+			valuesGranted = grants.addAllGrants(fqn, v.GetGrants(), def)
+			storeKeysToCache(v.GetGrants(), keyCache)
+		}
+		// If no more specific grant was found, then add the value grants
+		if !valuesGranted {
+			attributesGranted = grants.addAllGrants(fqn, def.GetGrants(), def)
+			storeKeysToCache(def.GetGrants(), keyCache)
+		}
+		if !valuesGranted && !attributesGranted {
+			grants.addAllGrants(fqn, namespace.GetGrants(), def)
+			storeKeysToCache(namespace.GetGrants(), keyCache)
+		}
```

Comment:
Maybe simpler, if you don't mind using continue
```suggestion
		storeKeysToCache(v.GetGrants(), keyCache)
		if grants.addAllGrants(fqn, v.GetGrants(), def) {
			continue
		}

		// If no more specific grant was found, then add the value grants
		storeKeysToCache(def.GetGrants(), keyCache)
		if grants.addAllGrants(fqn, def.GetGrants(), def) {
			continue
		}
		storeKeysToCache(namespace.GetGrants(), keyCache)
		grants.addAllGrants(fqn, namespace.GetGrants(), def)
```